craft_parts/plugins/ant_plugin.py

```diff
@@ -0,0 +1,183 @@
+        ):
+            raise errors.PluginEnvironmentValidationError(
+                part_name=self._part_name,
+                reason=f"invalid ant version {version!r}",
+            )
+
+
+class AntPlugin(Plugin):
+    """A plugin for Apache Ant projects.
+
+    The plugin requires the ``ant`` tool installed on the system. This can
+    be achieved by adding the appropriate declarations to ``build-packages``
+    or ``build-snaps``, or by having it installed or built in a different
+    part. In this case, the name of the part supplying ``ant`` must be
+    "ant-deps".
+
+    Additionally, Java projects need a dev kit (jdk) to build and a runtime
+    environment (jre) to run. There are multiple choices here, but frequently
+    adding ``default-jdk-headless`` to ``build-packages`` and
+    ``default-jre-headless`` to ``stage-packages`` is enough.
+
+    Once built, the plugin will create the following structure in the part's
+    install dir (which will later be staged/primed/packaged):
+
+    - A ``bin/java`` symlink pointing to the actual ``java`` binary provided
+      by the jre;
+    - A ``jar/`` directory containing the .jar files generated by the build.
+
+    The ant plugin uses the common plugin keywords, plus the following ant-
+    specific keywords:
+
+    - ``ant-build-targets``
+      (list of strings)
+      The ant targets to build. These are directly passed to the ``ant``
+      command line.
+    - ``ant-build-file``
+      (str)
+      The name of the main ant build file. Defaults to ``build.xml``.
+    - ``ant-properties``
+      (dict of strings to strings)
+      A series of key: value pairs that are passed to ant as properties
+      (using the ``-D{key}={value}`` notation).
+    """
+
+    properties_class = AntPluginProperties
+    validator_class = AntPluginEnvironmentValidator
+
+    def get_build_snaps(self) -> Set[str]:
+        """Return a set of required snaps to install in the build environment."""
+        return set()
+
+    def get_build_packages(self) -> Set[str]:
+        """Return a set of required packages to install in the build environment."""
+        return set()
+
+    def get_build_environment(self) -> Dict[str, str]:
+        """Return a dictionary with the environment to use in the build step."""
+        return {}
+
+    def get_build_commands(self) -> List[str]:
+        """Return a list of commands to run during the build step."""
+        options = cast(AntPluginProperties, self._options)
+
+        command = ["ant"]
+
+        if options.ant_build_file:
+            command.extend(["-f", options.ant_build_file])
+
+        for prop_name, prop_value in options.ant_properties.items():
+            command.append(f"-D{prop_name}={prop_value}")
+
+        command.extend(options.ant_build_targets)
+
+        return [
+            " ".join(command),
+        ] + get_java_post_build_commands()
+
+
+def get_java_post_build_commands() -> List[str]:
```

Comment:
This is what the core18 ant and maven plugins do; as a function here we can reuse it later for other java-based plugins

Reply:
Maybe we could have a Java base plugin superclass providing the infrastructure as protected methods to java-related plugins? We can think about that and refactor later if we decide to do something like that.